File: cpp/linked-lists/0002-add-two-numbers.cpp
#include <structures.h>

ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
    return addLists(l1, l2, 0);
}

ListNode *addLists(ListNode *l1, ListNode *l2, int carry) {
    if (l1 == nullptr && l2 == nullptr && carry == 0) {
        return nullptr;
    }
    
    int val = carry;
    if (l1 != nullptr) {
        val += l1->val;
    }
    
    if (l2 != nullptr) {
        val += l2->val;
    }
    
    ListNode *result = new ListNode(val % 10);
    
    result->next = addLists(l1 != nullptr ? l1->next : nullptr,
                            l2 != nullptr ? l2->next : nullptr,
                            val > 9 ? 1 : 0);

    return result;
}

File: js/linked-lists/0002-add-two-numbers.js
var addTwoNumbers = function(l1, l2) {
    return add(l1, l2, 0);
};

var add = function(l1, l2, carry) {
    if (l1 === null && l2 === null && carry === 0) return null;
    
    var val = carry;
    if (l1 !== null) {
        val += l1.val;
    }
    
    if (l2 !== null) {
        val += l2.val;
    }
    
    return {
        val: val % 10,
        next: add(l1 !== null ? l1.next : null,
            l2 !== null ? l2.next : null,
            val > 9 ? 1 : 0
        )
    };
}